popde/adaptive_kde.py

```diff
@@ -0,0 +1,206 @@
+    sample3 = rndgen.normal(mean3, sigma3, size=n_samples)
+    sample = np.column_stack((sample1, sample2, sample3)) # shape is (n_points, n_features)
+    # Create and fit the adaptive KDE note backend must be set correctly
+    kde = AdaptiveBwKDE(sample, backend='awKDEpy', dim_names=['x', 'y', 'z'], alpha=0.5, input_transf=None)
+    # Generate grid for plotting
+    xgrid = np.linspace(sample1.min(), sample1.max(), 100)
+    ygrid = np.linspace(sample2.min(), sample2.max(), 100)
+    zgrid = np.linspace(sample3.min(), sample3.max(), 100)
+    XX, YY, ZZ = np.meshgrid(xgrid, ygrid, zgrid)
+    eval_pts = np.column_stack((XX.flatten(), YY.flatten(), ZZ.flatten()))
+    kde.fit()
+
+    # Evaluate the KDE at the grid points
+    density_values = kde.evaluate(eval_pts)
+    density_values = density_values.reshape(XX.shape)
+    print(density_values)
+    """
+    def __init__(self, data, backend='KDEpy', bandwidth=1., dim_names=None, alpha=0.5,  input_transf=None, stdize=False, rescale=None):
+        # Additional parameter for the Wang & Wang formula
+        self.alpha = alpha
+        super().__init__(data, input_transf, stdize, rescale,
+                         backend, bandwidth, dim_names)  
+
+    def _local_bandwidth_factor(self, kde_values):
+        """
+        Calculate local bandwidth factor using expression in
+        B. Wang and X. Wang, 2007, DOI: 10.1214/154957804100000000.
+
+        Parameters:
+        -----------
+        kde_values : array-like
+            The KDE values at the data point positions.
+
+        Returns:
+        --------
+        loc_bw_factor : array-like
+           local bandwidth factor for adaptive 
+           bandwidth.
+        """
+        from scipy.stats import gmean
+        #geometric mean of kde values
+        g = gmean(kde_values)
+        loc_bw_factor = (kde_values / g) ** self.alpha
+        return loc_bw_factor
+
+    def calculate_per_point_bandwidths(self, pilot_values):
+        """
+        Calculate per-point bandwidths using the Wang & Wang formula in
+        B. Wang and X. Wang, 2007, DOI: 10.1214/154957804100000000.
+        Parameters:
+        -----------
+        pilot_values : array-like
+            The pilot KDE values at the data point positions.
+
+        Returns:
+        --------
+        per_point_bandwidths : array-like
+            Per-point bandwidths calculated using the Wang & Wang formula.
+            bw_i =  global_bw * local_bw_factor
+
+                local_bw_factor = (f(X_i)/g)^alpha
+                f(X_i) is pilot-KDE values
+                g is local factor given in local_bandwidth_factor
+        """
+        # Calculate the inverse of local bandwidth using the provided function
+        local_bandwidths = self._local_bandwidth_factor(pilot_values)
+
+        # Use the local bandwidths to calculate per-point bandwidths
+        self.bandwidth = self.bandwidth / local_bandwidths
+
+
+    def fit_awKDEpy(self):
```

Comment:
yes, you do not need this method, the adaptive kde calculation of bandwidth can be done within the __ini__ method of this class and then you can use the reassign_bandwidth method of the parent class to set the loca bandwidth and redo the fit.

Reply:
removed this and added adaptive bandwidth in init.